File: packages/chain/src/chain/bitcoin.ts
/**
 * An 80-byte Bitcoin block header, and the two things it can be asked offline.
 *
 * This is the whole of what a witness costs a reader. A timestamp proof folds a
 * digest into a block's MERKLE ROOT; the header is where that root is written,
 * next to the instant the block claims and the difficulty it was mined at. So a
 * record that carries the header carries everything needed to check the
 * attestation with no network, no node, and no service — which is the only reason
 * `verify` can stay offline while gaining a layer that is, by definition, outside
 * this machine.
 *
 * TWO QUESTIONS AND NOT THREE. This module answers whether the header CARRIES WORK
 * ({@link headerCarriesRealWork}) and what it says (merkle root, instant). It does
 * NOT answer whether the header sits on the longest chain — that needs the headers
 * before it, which this record does not carry and this package will not pretend to.
 *
 * SELF-CONSISTENCY IS NOT ENOUGH AND THE FLOOR IS WHY THIS IS NOT A ONE-LINER. A
 * header declares the difficulty it was mined at, so a header checked only against
 * its OWN declaration is free to forge: declare the easiest target there is, find a
 * nonce in milliseconds, and the check passes over a block nobody mined. So the
 * declared target is held to {@link MINIMUM_WORK} as well, and that turns a
 * millisecond into the work of a real block.
 *
 * WHAT IS LEFT, said here rather than in a footnote: the floor bounds the COST of a
 * forgery, it does not prove the block is Bitcoin's. Someone with mining hardware
 * and a reason could produce a header that passes this and is on no chain at all.
 * Closing that needs the chain of headers or somebody else's copy of it — and the
 * `.ots` beside the header is exactly what makes that a five-second job for a
 * stranger, who runs the OpenTimestamps client against a node and gets the answer
 * this module deliberately does not fake.
 *
 * EVERY FIELD IS LITTLE-ENDIAN and the merkle root is in INTERNAL byte order —
 * the order it is serialized in, which is the reverse of the order an explorer
 * prints. That matters exactly once: an OpenTimestamps Bitcoin attestation folds
 * to the root in internal order, so the comparison is against these bytes as they
 * lie, never against a display form.
 */

import { createHash } from 'node:crypto';

/** A header is exactly this many bytes; anything else is not one. */
export const BLOCK_HEADER_BYTES = 80;

/**
 * The weakest block this package will accept as a witness, in the compact form a
 * header declares: `0x1800ffff`, which is 2**40 times the difficulty of the genesis
 * block.
 *
 * WHY A CONSTANT AND NOT THE CURRENT DIFFICULTY. Nothing offline knows what the
 * current difficulty is — that is a fact about a chain this record does not carry —
 * so the choice is between a fixed floor and none. A fixed floor ages in ONE
 * direction that matters: it is a refusal, so it can only ever reject a real block
 * mined below it, never accept a forged one above it.
 *
 * WHY THIS NUMBER. It is roughly 2**72 hashes for one header, which is hours of the
 * entire Bitcoin network or years of a large private one — expensive enough that a
 * forged sidecar is not a thing somebody does casually, which is what the guard is
 * for. Bitcoin passed it in 2019 and would have to fall more than a hundredfold to
 * come back under it; if it ever does, this constant is the one line to revisit, and
 * the refusal it produces is loud rather than silent.
 */
export const MINIMUM_WORK = 0x1800ffff;

const MERKLE_ROOT_AT = 36;
const TIME_AT = 68;
const BITS_AT = 72;

/** Refused bytes: not a block header, or one that does not meet its own claim. */
export class UnreadableHeaderError extends Error {
  constructor(reason: string) {
    super(`block header: ${reason}`);
    this.name = 'UnreadableHeaderError';
  }
}

/** What an 80-byte header says. */
export interface BlockHeader {
  /** The header's own bytes, kept so the record can be re-checked from them. */
  readonly bytes: Buffer;
  /** The merkle root in INTERNAL byte order — what an attestation folds to. */
  readonly merkleRoot: Buffer;
  /** The instant the block claims, in seconds since the epoch. */
  readonly time: number;
  /** The compact difficulty target the block declares. */
  readonly bits: number;
  /** The block's hash, written the way an explorer prints it (reversed). */
  readonly id: string;
}

/** Reads an 80-byte header. */
export function parseBlockHeader(bytes: Buffer): BlockHeader {
  if (bytes.length !== BLOCK_HEADER_BYTES) {
    throw new UnreadableHeaderError(`${bytes.length} bytes, not ${BLOCK_HEADER_BYTES}`);
  }
  const hash = sha256d(bytes);
  return {
    bytes: Buffer.from(bytes),
    merkleRoot: Buffer.from(bytes.subarray(MERKLE_ROOT_AT, TIME_AT)),
    time: bytes.readUInt32LE(TIME_AT),
    bits: bytes.readUInt32LE(BITS_AT),
    id: Buffer.from(hash).reverse().toString('hex'),
  };
}

/**
 * Whether the header carries real work: its own hash under its own declared target,
 * AND that target no weaker than {@link MINIMUM_WORK}.
 *
 * BOTH HALVES OR NEITHER. The first alone accepts a header mined in a millisecond at
 * a declared difficulty of nothing; the second alone accepts a header that declares a
 * hard target and never met it. The mutation that proves each is the other's absence,
 * and they redden different cases (`bitcoin.test.ts`).
 *
 * The target is unpacked from the compact `bits` field exactly as a node unpacks it,
 * and the hash is compared as a 256-bit little-endian number, which is what the hash
 * IS: the leading zeroes an explorer shows are the trailing bytes here.
 *
 * A `bits` field a header could not have carried — the sign bit set, an exponent past
 * the width of the number, a zero mantissa — is FALSE rather than an exception: it is
 * a header that fails its own claim, which is the same answer as a nonce that never
 * found one, and a caller has one thing to do about either.
 */
export function headerCarriesRealWork(header: BlockHeader): boolean {
  const target = targetOf(header.bits);
  const floor = targetOf(MINIMUM_WORK);
  if (target === null || floor === null) return false;
  // A LARGER target is an EASIER block, so the floor is an upper bound on the target.
  if (target.compare(floor) > 0) return false;
  const hash = Buffer.from(sha256d(header.bytes)).reverse();
  return hash.compare(target) <= 0;
}

/** The 256-bit target a compact `bits` field means, or null if it means none. */
function targetOf(bits: number): Buffer | null {
  const exponent = bits >>> 24;
  const mantissa = bits & 0x007fffff;
  // The sign bit of the compact form: a header that set it declares a negative
  // target, which no block can meet.
  if ((bits & 0x00800000) !== 0) return null;
  if (mantissa === 0 || exponent > 32 || exponent < 3) return null;
  const target = Buffer.alloc(32);
  // The mantissa's three bytes sit so that their LAST one is at position `exponent`,
  // counting from the big end — the compact form's definition.
  target.writeUIntBE(mantissa, 32 - exponent, 3);
  return target;
}

/** SHA-256 twice, which is how Bitcoin hashes a header. */
function sha256d(bytes: Buffer): Buffer {
  return createHash('sha256').update(createHash('sha256').update(bytes).digest()).digest();
}

File: packages/chain/verifier/mnemaverify/bitcoin.py
"""Section 8's arithmetic on a Bitcoin block header.

An 80-byte header, in the order Bitcoin serializes it:

    0..4    version              little-endian
    4..36   previous block hash
    36..68  the merkle root      "internal order" - the bytes as stored, which is the
                                 reverse of how an explorer prints it
    68..72  time                 little-endian
    72..76  bits                 the compact target, little-endian
    76..80  nonce

Two of section 8's three questions are here: the path folds to the merkle root the stored
header carries, and the header's own hash meets the target it declares.

GAP G17. Section 8 names the endianness for the merkle root ("bytes 36-68, internal
order") and names nothing for the target: not the compact nBits expansion, not that the
double-SHA256 is compared as a little-endian integer. Both are Bitcoin standard, so a
stranger gets there - but the document chose to spell out one of the two and not the other.

The floor checks out as arithmetic: 0x1800ffff expands to 0xffff * 2**168, the genesis
target 0x1d00ffff expands to 0xffff * 2**208, and 208 - 168 = 40. So "at least 0x1800ffff"
and "2**40 times the genesis difficulty" are the same sentence, as section 8 says.
"""

from __future__ import annotations

import hashlib
from typing import NamedTuple

from .verdict import Refusal

HEADER_LENGTH = 80
MERKLE_ROOT_SLICE = slice(36, 68)
TIME_SLICE = slice(68, 72)
BITS_SLICE = slice(72, 76)

# Section 8's declared floor. A header mined at an easier target is refused, because one
# is found in milliseconds.
MINIMUM_BITS = 0x1800FFFF
GENESIS_BITS = 0x1D00FFFF


def _sha256d(data: bytes) -> bytes:
    return hashlib.sha256(hashlib.sha256(data).digest()).digest()


def target_from_bits(bits: int) -> int:
    exponent = bits >> 24
    mantissa = bits & 0x007FFFFF
    if bits & 0x00800000:  # the sign bit; no real header sets it
        raise Refusal("8", f"a negative compact target: {bits:#010x}")
    if exponent <= 3:
        return mantissa >> (8 * (3 - exponent))
    return mantissa << (8 * (exponent - 3))


MINIMUM_TARGET = target_from_bits(MINIMUM_BITS)


class Header(NamedTuple):
    raw: bytes
    height: int
    merkle_root: bytes
    stamped_at: int
    bits: int
    block_id: str

    @property
    def instant(self) -> str:
        """The header's own timestamp, as section 8's "the instant"."""
        from datetime import datetime, timezone

        return datetime.fromtimestamp(self.stamped_at, tz=timezone.utc).isoformat(
            timespec="seconds"
        )

    @property
    def work_hash(self) -> int:
        return int.from_bytes(_sha256d(self.raw), "little")

    @property
    def target(self) -> int:
        return target_from_bits(self.bits)

    def meets_its_own_target(self) -> bool:
        return self.work_hash <= self.target

    def meets_the_declared_floor(self) -> bool:
        return self.target <= MINIMUM_TARGET


def parse_header(raw: bytes, height: int) -> Header:
    if len(raw) != HEADER_LENGTH:
        raise Refusal("8", f"a block header of {len(raw)} bytes, not {HEADER_LENGTH}")
    return Header(
        raw=raw,
        height=height,
        merkle_root=raw[MERKLE_ROOT_SLICE],
        stamped_at=int.from_bytes(raw[TIME_SLICE], "little"),
        bits=int.from_bytes(raw[BITS_SLICE], "little"),
        # The id an explorer prints: the double-SHA256 reversed. Reported so that section
        # 8's escape hatch - "follows the block id into any explorer" - is usable from here.
        block_id=_sha256d(raw)[::-1].hex(),
    )


def read_blocks_sidecar(raw: bytes) -> dict[int, Header]:
    """The `.blocks` file: one canonical JSON object per line, {"header","height"}.

    Gap G16: the sidecar may carry headers the proof beside it does not use - the frozen
    record has one `.ots` and two headers - and the document does not say that, nor that
    the pairing is by the height the attestation declares.
    """
    from .canonical import canonical_bytes, strict_loads

    headers: dict[int, Header] = {}
    for number, line in enumerate(raw.split(b"\n"), start=1):
        if not line.strip():
            continue
        stored = strict_loads(line.decode("utf-8"))
        if not isinstance(stored, dict) or set(stored) != {"header", "height"}:
            raise Refusal("8", f"a .blocks line whose keys are not header and height (line {number})")
        if canonical_bytes(stored) != line:
            raise Refusal("8", f"a .blocks line that is not canonical (line {number})")
        if not isinstance(stored["height"], int) or isinstance(stored["height"], bool):
            raise Refusal("8", f"a .blocks height that is not an integer (line {number})")
        hex_header = stored["header"]
        if not isinstance(hex_header, str) or len(hex_header) != HEADER_LENGTH * 2:
            raise Refusal("8", f"a .blocks header that is not 160 hex characters (line {number})")
        try:
            data = bytes.fromhex(hex_header)
        except ValueError as exc:
            raise Refusal("8", f"a .blocks header that is not hex (line {number})") from exc
        headers[stored["height"]] = parse_header(data, stored["height"])
    return headers
